File: vscode/src/tooltip_renderer.ts
// TooltipRenderer turns a Snapshot into the MarkdownString the VS Code
// status-bar item shows on hover.
import * as vscode from 'vscode';
import { Snapshot } from './snapshot';
import { formatCostUSD, formatTokensCompact } from './format';

export function renderTooltip(snapshot: Snapshot): vscode.MarkdownString {
  const window5h = snapshot.window5h;
  const weightedCacheRead = Math.round(window5h.cacheReadTokens * 0.1);

  const markdownTooltip = new vscode.MarkdownString();
  markdownTooltip.supportThemeIcons = true;
  markdownTooltip.isTrusted = false;

  markdownTooltip.appendMarkdown(
    `**Claude Hunter** · ${snapshot.model ?? 'unknown model'}\n\n`,
  );
  markdownTooltip.appendMarkdown(
    `**5h window** — ${formatTokensCompact(window5h.effectiveTokens)} effective ` +
      `(${window5h.percentOfCeilingEstimate.toFixed(1)}% of ceiling)\n\n`,
  );

  markdownTooltip.appendMarkdown('| | |\n|---|---|\n');
  markdownTooltip.appendMarkdown(
    `| Input | ${formatTokensCompact(window5h.inputTokens)} |\n` +
      `| Output | ${formatTokensCompact(window5h.outputTokens)} |\n` +
      `| Cache write | ${formatTokensCompact(window5h.cacheCreateTokens)} |\n` +
      `| Cache read | ${formatTokensCompact(window5h.cacheReadTokens)} ` +
      `(×0.1 → ${formatTokensCompact(weightedCacheRead)}) |\n` +
      `| Cost | ${formatCostUSD(window5h.costUSD)} |\n` +
      `| Burn rate | ${Math.round(window5h.burnRatePerMinute)} tok/min |\n`,
  );

  if (window5h.perModel && window5h.perModel.length > 0) {
    markdownTooltip.appendMarkdown('\n**By model**\n\n');
    markdownTooltip.appendMarkdown('| Model | Tokens | Cost |\n|---|---:|---:|\n');
    for (const modelEntry of window5h.perModel) {
      markdownTooltip.appendMarkdown(
        `| ${modelEntry.model} | ` +
          `${formatTokensCompact(modelEntry.effectiveTokens)} | ` +
          `${formatCostUSD(modelEntry.costUSD)} |\n`,
      );
    }
  }

  return markdownTooltip;
}

File: intellij/src/main/kotlin/co/reward360/claudehunter/TooltipRenderer.kt
package co.reward360.claudehunter

// TooltipRenderer turns a Snapshot into the HTML string an IntelliJ status
// bar tooltip needs (plain \n is ignored by Swing tooltips).
object TooltipRenderer {
    fun render(snapshot: Snapshot): String {
        val window = snapshot.window
        return buildString {
            append("<html><body style='padding:2px 6px;'>")
            append("<b>Claude Hunter</b> &middot; ")
            append(escapeHtml(snapshot.model ?: "unknown model"))
            append("<br><br>")

            append("<b>5h window</b> &mdash; ")
            append(SnapshotFormat.tokensCompact(window.effectiveTokens))
            append(" effective (")
            append("%.1f".format(window.percentOfCeilingEstimate))
            append("% of ceiling)")
            append("<br><br>")

            append("<table cellpadding='1' cellspacing='0'>")
            appendRow("Input", SnapshotFormat.tokensCompact(window.inputTokens))
            appendRow("Output", SnapshotFormat.tokensCompact(window.outputTokens))
            appendRow("Cache write", SnapshotFormat.tokensCompact(window.cacheCreateTokens))
            appendRow(
                "Cache read",
                "${SnapshotFormat.tokensCompact(window.cacheReadTokens)} " +
                    "(&times;0.1 &rarr; ${SnapshotFormat.tokensCompact((window.cacheReadTokens * 0.1).toLong())})",
            )
            appendRow("Cost", SnapshotFormat.costUSD(window.costUSD))
            appendRow("Burn rate", "${window.burnRatePerMinute.toInt()} tok/min")
            append("</table>")

            if (window.perModel.isNotEmpty()) {
                append("<br><b>By model</b><br>")
                append("<table cellpadding='1' cellspacing='0'>")
                append("<tr><td style='padding-right:12px; color:#888;'>Model</td>")
                append("<td style='padding-right:12px; color:#888;'>Tokens</td>")
                append("<td style='color:#888;'>Cost</td></tr>")
                for (modelEntry in window.perModel) {
                    append("<tr><td style='padding-right:12px;'>")
                    append(escapeHtml(modelEntry.model))
                    append("</td><td style='padding-right:12px;'>")
                    append(SnapshotFormat.tokensCompact(modelEntry.effectiveTokens))
                    append("</td><td>")
                    append(SnapshotFormat.costUSD(modelEntry.costUSD))
                    append("</td></tr>")
                }
                append("</table>")
            }

            append("</body></html>")
        }
    }

    private fun StringBuilder.appendRow(label: String, value: String) {
        append("<tr><td style='padding-right:12px; color:#888;'>")
        append(label)
        append("</td><td>")
        append(value)
        append("</td></tr>")
    }

    private fun escapeHtml(raw: String): String =
        raw.replace("&", "&amp;")
            .replace("<", "&lt;")
            .replace(">", "&gt;")
}
